dataset/shape.cpp

```diff
@@ -141,4 +144,190 @@ Dataset resize(const DatasetConstView &d, const Dim dim,
   return result;
 }
 
+namespace {
+
+/// Either broadcast variable to from_dims before a reshape or not:
+///
+/// 1. If all from_dims are contained in the variable's dims, no broadcast
+/// 2. If at least one (but not all) of the from_dims is contained in the
+///    variable's dims, broadcast
+/// 3. If none of the variables's dimensions are contained, no broadcast
+Variable maybe_broadcast(const VariableConstView &var,
+                         const Dimensions &from_dims) {
+  const auto &var_dims = var.dims();
+
+  Dimensions broadcast_dims;
+  for (const auto dim : var_dims.labels())
+    if (!from_dims.contains(dim))
+      broadcast_dims.addInner(dim, var_dims[dim]);
+    else
+      for (const auto lab : from_dims.labels())
+        if (!broadcast_dims.contains(lab)) {
+          // Need to check if the variable contains that dim, and use the
+          // variable shape in case we have a bin edge.
+          if (var_dims.contains(lab))
+            broadcast_dims.addInner(lab, var_dims[lab]);
+          else
+            broadcast_dims.addInner(lab, from_dims[lab]);
+        }
+  return broadcast(var, broadcast_dims);
+}
+
+/// Special handling for folding coord along a dim that contains bin edges.
+///
+/// The procedure is the following:
+/// - create output of correct size (use fold_dims and resize to_dims.inner()
+///   to L+1).
+/// - copy(reshape(coord.slice({from_dim, 0, size-1}), to_dims),
+///   out_coord.slice({to_dims.inner(), 0, Ninner})
+/// - copy(reshape(coord.slice({from_dim, 1, size}),
+///   to_dims).slice({to_dims.inner(), Ninner-1}),
+///   out_coord.slice({to_dims.inner(), Ninner})
```

Comment:
Maybe these comments are redundant since you have so many inline comments.

Reply:
I guess they were for doxygen?